File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    for i in permutations(dungeons):
        tmp = k
        cnt = 0
        for start, end in i:
            if tmp < start:
                break
            else:
                tmp -= end
                cnt += 1
        answer = max(answer, cnt)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
  var answer = -1;
  const visited = Array(dungeons.length).fill(false);

  function dfs(now, cnt) {
    answer = Math.max(answer, cnt);

    for (let i = 0; i < dungeons.length; i++) {
      const [start, end] = dungeons[i];

      if (!visited[i] && now >= start) {
        visited[i] = true;
        dfs(now - end, cnt + 1);
        visited[i] = false;
      }
    }
  }

  dfs(k, 0);
  return answer;
}
